### PR TITLE
Inject container status and task relocation data in the job event stream

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

**Review by amitaekbote — APPROVED** [review]

File: titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/GatewayJobServiceGateway.java [hunk 1]
```diff
@@ -294,17 +293,21 @@ public GatewayJobServiceGateway(GrpcRequestConfiguration tunablesConfiguration,
     @Override
     public Observable<JobChangeNotification> observeJob(String jobId, CallMetadata callMetadata) {
         if (localCacheQueryProcessor.canUseCache(Collections.emptyMap(), "observeJob", callMetadata)) {
-            return localCacheQueryProcessor.syncCache("observeJob", JobChangeNotification.class).concatWith(localCacheQueryProcessor.observeJob(jobId));
+            return localCacheQueryProcessor.syncCache("observeJob", JobChangeNotification.class)
+                    .concatWith(localCacheQueryProcessor.observeJob(jobId))
+                    .map(taskRelocationDataInjector::injectIntoTaskUpdateEvent);
         }
-        return super.observeJob(jobId, callMetadata);
+        return super.observeJob(jobId, callMetadata).map(taskRelocationDataInjector::injectIntoTaskUpdateEvent);
```
Comment (solarkennedy) [inline titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/GatewayJobServiceGateway.java:300]: I've not seen this code before, but before, this localCacheQueryProcessor thing was doing the heavy lifting, regardless of the number of observeJobs streamers.

But now with the `.map()`, will we be calling (or really, looking at the shared informer) for each job and for each streamer? If so, any worries about scaling here?

**Review by solarkennedy — APPROVED** [review]

File: titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/GatewayJobServiceGateway.java (hunk 1 above)
Comment (tbak) [reply titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/GatewayJobServiceGateway.java:300]: `localCacheQueryProcessor` provides the raw data. The data injector decorates them with additional information.
The shared informer calls should be inexpensive, just like setting the data in the protobuf messages.

File: titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/TaskDataInjector.java [hunk 2]
```diff
@@ -182,11 +205,11 @@ private long getTaskRelocationTimeout() {
 
     private Task newTaskWithContainerState(Task task) {
         return task.toBuilder().setStatus(TaskStatus.newBuilder()
-                    .addAllContainerState(kubeApiConnector.getContainerState(task.getId()))).build();
+                .addAllContainerState(kubeApiConnector.getContainerState(task.getId()))).build();
```
Comment (solarkennedy) [inline titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/TaskDataInjector.java:208]: I think this line nukes out whatever original status stuff we had when we do `setStatus` to this new status we build! (which only has ContainerState, and none of whatever we originally had)

Comment (solarkennedy) [reply titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/TaskDataInjector.java:208]: Should be `task.toBuilder().setStatus(task.getStatus().toBuilder()` I think

Comment (amitaekbote) [reply titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/TaskDataInjector.java:208]: https://github.com/Netflix/titus-control-plane/pull/1189

**Pull request merged**